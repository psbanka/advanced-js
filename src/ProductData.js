import React, {Component} from 'react'
import ProductRow from './ProductRow'

export default class ProductData extends Component {

  _generateTableGuts () {
    let tableRow = []
    let categories = []
    this.props.catalog.forEach((catalogEntry) => {
      const newItem = categories.indexOf(catalogEntry.category) === -1
      if (newItem) {
        categories.push(catalogEntry.category)
      }
    })

    categories.forEach((currentCategory) => {
      tableRow.push(
        <tr key={currentCategory}>
          <td colSpan='2'><strong>{currentCategory}</strong></td>
        </tr>
      )
      this.props.catalog.forEach((catalogEntry) => {
        let key = `${currentCategory}${catalogEntry.name}`
        if (currentCategory === catalogEntry.category) {
          tableRow.push(
            <ProductRow
              key={key}
              {...catalogEntry}
              currentCategory={currentCategory}
              searchTerm={this.props.searchTerm}
              inStock={this.props.inStock}
              isBuying={this.props.isBuying}
              onIsBuying={this.props.onIsBuying}
            />
          )
        }

      })
    })
    return tableRow
  }

  render () {
    const tableRow = this._generateTableGuts()
    return (
      <tbody>
        {tableRow}
      </tbody>
    )
  }
}


