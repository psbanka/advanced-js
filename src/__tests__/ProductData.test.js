import React from 'react'
import ReactDOM from 'react-dom'
import ProductData from '../ProductData'
import { shallow, mount, render } from 'enzyme'
import renderer from 'react-test-renderer';

/* global it describe */

const CATALOG = [
  {category: 'things', 'price': 99.99, stocked: false, name: 'bigThing'},
  {category: 'things', 'price': 9.99, stocked: true, name: 'littleThing'},
  {category: 'stuff', 'price': 1.99, stocked: true, name: 'junkyThing'}
]

describe('ProductData', () => {
  it('snapshot check', () => {
    const component = renderer.create(
      <ProductData
        catalog={CATALOG}
        searchTerm=''
        inStock={false}
        isBuying={{}}
        onIsBuying={{}}
      />
    )
    const json = component.toJSON()
    expect(json).toMatchSnapshot()
  })

  it('renders table', () => {
    const wrapper = shallow(
      <ProductData
        catalog={CATALOG}
        searchTerm=''
        inStock={false}
        isBuying={{}}
        onIsBuying={{}}
      />
    )
    const output = wrapper.instance()._generateTableGuts()
    const expectedKeys = [
      'things',
      'thingsbigThing',
      'thingslittleThing',
      'stuff',
      'stuffjunkyThing'
    ]
    expectedKeys.forEach((key, index) => {
      expect(output[index].key).toBe(key)
    })
  })
})
