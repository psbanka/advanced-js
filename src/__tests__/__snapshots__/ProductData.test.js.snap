// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ProductData snapshot check 1`] = `
<tbody>
  <tr>
    <td
      colSpan="2"
    >
      <strong>
        things
      </strong>
    </td>
  </tr>
  <tr
    className="product"
    style={
      Object {
        "color": "red",
      }
    }
  >
    <td>
      <input
        checked={false}
        id="things-bigthing"
        onChange={[Function]}
        type="checkbox"
      />
      bigThing
    </td>
    <td>
      $
      99.99
    </td>
  </tr>
  <tr
    className="product"
    style={
      Object {
        "color": "black",
      }
    }
  >
    <td>
      <input
        checked={false}
        id="things-littlething"
        onChange={[Function]}
        type="checkbox"
      />
      littleThing
    </td>
    <td>
      $
      9.99
    </td>
  </tr>
  <tr>
    <td
      colSpan="2"
    >
      <strong>
        stuff
      </strong>
    </td>
  </tr>
  <tr
    className="product"
    style={
      Object {
        "color": "black",
      }
    }
  >
    <td>
      <input
        checked={false}
        id="stuff-junkything"
        onChange={[Function]}
        type="checkbox"
      />
      junkyThing
    </td>
    <td>
      $
      1.99
    </td>
  </tr>
</tbody>
`;
